verus/lcf/README.md

```diff
@@ -0,0 +1,396 @@
+7. pub open spec fn spec_valid(self) -> bool
+8. pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecProp)
+9. pub fn prop_eq(&self, other: &Self) -> (res: bool)
+10. pub fn valid(self) -> (res: bool)
+11. pub fn symbolic(self) -> (res: bool)
+12. pub fn concrete(self) -> (res: bool)
+13. pub fn complete_subst(&self, s: &Subst) -> (res: bool)
+14. pub fn subst(&self, s: &Subst) -> (res: Prop)
+
+## Rule
+
+Rules are the equivalent to rules in Datalog. A Rule in the kernel with no body is equivalent to a fact in Datalog. In this datalog line of code: connected(a, b) :- edge(a, b). ---- connected(a, b) :- edge(a, b) is a Rule that is also a rule in Datalog. In this datalog line of code: edge("x", "y"). ---- edge("x", "y"). is a Rule that is a fact in Datalog because it only has a head and no body.
+
+Rule types are structs that have 3 members:
+1.  pub head: Prop
+2. pub body: Vec<Prop>
+3. pub id: u64
+
+Its spec versions for SpecRule are:
+1. pub head: SpecProp
+2. pub body: Seq<SpecProp>
+3. pub id: u64
+
+Functions that can be used on Rule types:
+1. (in impl DeepView for Rule) open spec fn deep_view(&self) -> Self::V
+2. (in impl Clone for Rule) fn clone(&self) -> (res: Self)
+3. (in impl PartialEq for Rule) fn eq(&self, other: &Self) -> (res: bool)
+4. pub open spec fn spec_complete_subst(self, s: SpecSubst) -> bool
+5. pub open spec fn spec_concrete(self) -> bool
+6. pub open spec fn spec_wf(self) -> bool
+7. pub open spec fn spec_subst(self, s: SpecSubst) -> (res: SpecRule)
+8. pub fn subst(&self, s: &Subst) -> (res: Rule)
+9. pub fn complete_subst(&self, s: &Subst) -> (res: bool)
+10. pub fn concrete(&self) -> (res: bool)
+11. pub fn wf(&self) -> (res: bool)
+
+## Ruleset
+
+A RuleSet is a set of Rules in our kernel. 
+Ruleset types are structs that have 1 member:
+1. pub rs: Vec<Rule>
+
+Its spec version for SpecRule is:
+1. pub rs: Seq<SpecRule>
+
+Functions that can be used on RuleSet types:
+1. (in impl DeepView for RuleSet) open spec fn deep_view(&self) -> Self::V
+2. pub open spec fn spec_wf(self) -> bool
+3. pub fn wf(self) -> (res: bool)
+
+## Proof
+
+Proofs are used to construct thms in our kernel. 
+Proof types are enums with 2 members:
+1. Pstep(Rule, Subst, Vec<Proof>)
+2. QED(Prop)
+
+Its spec versions for SpecProof are:
+1. Pstep(SpecRule, SpecSubst, Seq<SpecProof>)
+2. QED(SpecProp)
+
+Functions that can be used on Proof types:
+1. (in impl DeepView for Proof) open spec fn deep_view(&self) -> Self::V
+2. (in impl Clone for Proof) fn clone(&self) -> (res: Self)
+3. (in impl PartialEq for Proof) fn eq(&self, other: &Self) -> (res: bool)
+4. pub open spec fn spec_valid(self, rule_set: SpecRuleSet) -> bool
+5. pub open spec fn spec_head(self) -> SpecProp
+6. pub fn head(&self) -> (res: Prop)
+
+
+## Thm
+Theorems (thm) are used in our kernel as the final result that we are producing and want to run Ok on. If the theorem is valid, Ok will be returned, and if it is invalid, Err will be returned.
```

Comment:
Not sure what you mean by "run Ok on" here. The `Ok` is a constructor from the `Result` type, which is used in `mk_thm` to tell us whether we successfully constructed a `Thm` or not. The key thing about `Thm` is that if it is well-formed, then we know that there exists a valid proof tree for the atom denoted by `val`.